rmc/src/lib.rs

```diff
@@ -0,0 +1,748 @@
+                Ok(multisigned) => {
+                    self.on_complete_multisignature(multisigned);
+                }
+                Err(_) => {
+                    warn!(target: "AlephBFT-rmc", "Received a hash with a bad multisignature");
+                }
+            },
+            Message::SignedHash(unchecked) => {
+                let signed_hash = match unchecked.check(self.keychain) {
+                    Ok(signed_hash) => signed_hash,
+                    Err(_) => {
+                        warn!(target: "AlephBFT-rmc", "Received a hash with a bad signature");
+                        return;
+                    }
+                };
+
+                let new_state = match self.hash_states.remove(&hash) {
+                    None => signed_hash.into_partially_multisigned(self.keychain),
+                    Some(partial) => partial.add_signature(signed_hash, self.keychain),
+                };
+                match new_state {
+                    PartiallyMultisigned::Complete { multisigned } => {
+                        self.on_complete_multisignature(multisigned)
+                    }
+                    incomplete => {
+                        self.hash_states.insert(hash.clone(), incomplete);
+                    }
+                }
+            }
+        }
+    }
+
+    fn do_task(&self, task: Task<H, MK>) {
+        let Task::BroadcastMessage(message) = task;
+        self.network_tx
+            .unbounded_send(message)
+            .expect("Sending message should succeed");
+    }
+
+    /// Fetches final multisignature.
+    pub fn get_multisigned(&self, hash: &H) -> Option<Multisigned<'a, H, MK>> {
+        match self.hash_states.get(hash)? {
+            PartiallyMultisigned::Complete { multisigned } => Some(multisigned.clone()),
+            _ => None,
+        }
+    }
+
+    /// Perform underlying tasks until the multisignature for the hash of this instance is collected.
+    pub async fn next_multisigned_hash(&mut self) -> Multisigned<'a, H, MK> {
+        loop {
+            futures::select! {
+                multisigned_hash = self.multisigned_hashes_rx.next() => {
+                    return multisigned_hash.expect("We own the tx, so it is not closed");
+                }
+
+                incoming_message = self.network_rx.next() => {
+                    if let Some(incoming_message) = incoming_message {
+                        self.handle_message(incoming_message);
+                    } else {
+                        debug!(target: "AlephBFT-rmc", "Network connection closed");
+                    }
+                }
+
+                task = self.scheduler.next_task().fuse() => {
+                    if let Some(task) = task {
+                        self.do_task(task);
+                    } else {
+                        debug!(target: "AlephBFT-rmc", "Tasks ended");
+                    }
+                }
+            }
+        }
+    }
+}
+
+#[cfg(test)]
```

Comment:
Perhaps move the tests to a separate file, this one is pretty long already and the tests are not quite unit tests anyway.